File: android/src/main/java/au/com/crazybean/mobilex/kurir/impl/FirebaseStorage.kt
package au.com.crazybean.mobilex.kurir.impl

import au.com.crazybean.mobilex.foundation.logger.Logger
import au.com.crazybean.mobilex.kurir.storage.CloudStorage
import com.google.firebase.firestore.DocumentChange
import com.google.firebase.firestore.FirebaseFirestore

class FirebaseStorage : CloudStorage {
    private val firestore by lazy {
        FirebaseFirestore.getInstance()
    }

    override fun readData(paths: String, completion: (Map<String, Any?>?, Throwable?) -> Unit) {
        firestore.document(paths)
            .get()
            .addOnSuccessListener { snapshot ->
                completion(snapshot.data, null)
            }
            .addOnFailureListener {
                completion(null, it)
            }
    }

    override fun writeData(paths: String, payload: Map<String, Any?>, completion: (Boolean, Throwable?) -> Unit) {
        firestore.document(paths)
            .set(payload)
            .addOnSuccessListener {
                completion(true, null)
            }
            .addOnFailureListener {
                completion(false, it)
            }
    }

    override fun delete(paths: String, completion: (Boolean, Throwable?) -> Unit) {
        firestore.document(paths)
            .delete()
            .addOnSuccessListener {
                completion(true, null)
            }
            .addOnFailureListener {
                completion(false, null)
            }
    }

    override fun readArray(paths: String,
                           completion: (List<Map<String, Any?>>?, Throwable?) -> Unit,
                           observation: ((List<Map<String, Any?>>?, List<Map<String, Any?>>?, List<Map<String, Any?>>?, Throwable?) -> Unit)?) {
        firestore.collection(paths).also { reference ->
            observation?.let { callback ->
                reference.addSnapshotListener { snapshot, exception ->
                    snapshot?.documentChanges?.takeIf { it.isNotEmpty() }?.let { changes ->
                        val added = changes.filter { it.type == DocumentChange.Type.ADDED }.map {
                            it.document.data
                        }

                        val modified = changes.filter { it.type == DocumentChange.Type.MODIFIED }.map {
                            it.document.data
                        }

                        val removed = changes.filter { it.type == DocumentChange.Type.REMOVED }.map {
                            it.document.data
                        }

                        callback(added, modified, removed, exception)
                    }?: Logger.d(exception)
                }
            }
        }.get()
            .addOnSuccessListener {
                val entities = it.map { document ->
                    document.data
                }
                completion(entities, null)
            }
            .addOnFailureListener {
                completion(null, it)
            }
    }
}

File: ios/Kurir/impl/FirebaseStorage.swift
//
//  DatabaseImpl.swift
//  Kurir
//
//  Created by Loren on 1/9/19.
//  Copyright © 2019 Crazybean Studio. All rights reserved.
//

import Mobilex
import Firebase

class FirebaseStorage: CloudStorage {
    private lazy var firestore: Firestore = {
        let firestore = Firestore.firestore()
        let settings = firestore.settings
        settings.areTimestampsInSnapshotsEnabled = true
        firestore.settings = settings
        return firestore
    }()
    
    func readData(paths: String, completion: @escaping ([String : Any]?, KotlinThrowable?) -> Void) {
        firestore.document(paths)
            .getDocument { snapshot, error in
                completion(snapshot?.data(), error?.throwable)
        }
    }
    
    func writeData(paths: String, payload: [String : Any], completion: @escaping (KotlinBoolean, KotlinThrowable?) -> Void) {
        firestore.document(paths)
            .setData(payload) { error in
                completion((error == nil).boolean, error?.throwable)
        }
    }
    
    func delete(paths: String, completion: @escaping (KotlinBoolean, KotlinThrowable?) -> Void) {
        firestore.document(paths)
            .delete { error in
                completion((error == nil).boolean, error?.throwable)
        }
    }
    
    func readArray(paths: String, completion: @escaping ([[String : Any]]?, KotlinThrowable?) -> Void,
                   observation: (([[String : Any]]?, [[String : Any]]?, [[String : Any]]?, KotlinThrowable?) -> Void)? = nil) {
        firestore.collection(paths)
            .getDocuments { snapshot, error in
                if let snapshot = snapshot {
                    completion(snapshot.documents.map{ $0.data() }, error?.throwable)
                } else {
                    completion(nil, error?.throwable)
                }
        }
    }
}

extension Error {
    var throwable: KotlinThrowable {
        get {
            return KotlinThrowable(message: localizedDescription)
        }
    }
}

extension Bool {
    var boolean: KotlinBoolean {
        get {
            return KotlinBoolean(value: self)
        }
    }
}

